# Feature Selection

--------------

Filtering Method

In [1]:
#Import all the library
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

-------

Constant Feature Removal

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.feature_selection import VarianceThreshold

In [3]:
cd /Users/apple/Downloads/Data-Files-for-Feature-Selection-master

/Users/apple/Downloads/Data-Files-for-Feature-Selection-master


In [4]:
pwd

'/Users/apple/Downloads/Data-Files-for-Feature-Selection-master'

-----

Load the Dataset

In [5]:
data = pd.read_csv("santander-train.csv")

In [6]:
data.head(10)

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   
5  13     2     23                 0.0                      0.0   
6  14     2     27                 0.0                      0.0   
7  18     2     26                 0.0                      0.0   
8  20     2     45                 0.0                      0.0   
9  23     2     25                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   
5                      0.0                      0.0                      0.0   
6                      0.0                      0.0                      0.0   
7                      0.0                      0.0                      0.0   
8                      0.0                      0.0                      0.0   
9                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   
2                      0.0                      0.0  ...   
3                      0.0                      0.0  ...   
4                      0.0                      0.0  ...   
5                      0.0                      0.0  ...   
6                      0.0                      0.0  ...   
7                      0.0                      0.0  ...   
8                      0.0                      0.0  ...   
9                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   
5                      0.0                      0.0                     0.0   
6                      0.0                      0.0                     0.0   
7                      0.0                      0.0                     0.0   
8                      0.0                      0.0                     0.0   
9                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   
5                     0.0                      0.0                      0.0   
6                     0.0                      0.0                      0.0   
7                     0.0                      0.0                      0.0   
8                     0.0                      0.0                      0.0   
9                     0.0                      0.0             

In [7]:
x = data.drop('TARGET', axis=1)
y=data['TARGET']

In [8]:
x.shape

(76020, 370)

In [9]:
y.shape

(76020,)

In [10]:
x_train, x_test, y_train, y_test = train_test_split (x , y, test_size=0.2, random_state=0)

-------------

Variance Threshhold

#In case we have lot of feature in dataset, we have to discard some feature to avoid curse of Dimensionality and to improve the accuracy of the prediction. Variance Threshold is used there to remove the features having almost same feature or ratherwe will say features with variances lower than threshhold

# Constant Removal

In [11]:
constant_filter = VarianceThreshold(threshold=0)
constant_filter.fit(x_train)

VarianceThreshold(threshold=0)

In [12]:
constant_filter.get_support().sum()

332

In [13]:
constant_list = [not temp for temp in constant_filter.get_support()]
constant_list

[False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 True,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 True,
 False,
 False,
 False,
 True,
 True,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 False,
 

In [14]:
x_train_filter = constant_filter.transform(x_train)
x_test_filter = constant_filter.transform(x_test)

In [15]:
x_train_filter.shape, x_test_filter.shape, x_train_filter.shape

((60816, 332), (15204, 332), (60816, 332))

In [16]:
#Check the number of features eleiminated
x.columns[constant_list]

Index(['ind_var2_0', 'ind_var2', 'ind_var27_0', 'ind_var28_0', 'ind_var28',
       'ind_var27', 'ind_var41', 'ind_var46_0', 'ind_var46', 'num_var27_0',
       'num_var28_0', 'num_var28', 'num_var27', 'num_var41', 'num_var46_0',
       'num_var46', 'saldo_var28', 'saldo_var27', 'saldo_var41', 'saldo_var46',
       'delta_imp_reemb_var33_1y3', 'delta_num_reemb_var33_1y3',
       'imp_amort_var18_hace3', 'imp_amort_var34_hace3',
       'imp_reemb_var13_hace3', 'imp_reemb_var33_hace3',
       'imp_reemb_var33_ult1', 'imp_trasp_var17_out_hace3',
       'imp_trasp_var33_out_hace3', 'num_var2_0_ult1', 'num_var2_ult1',
       'num_reemb_var13_hace3', 'num_reemb_var33_hace3',
       'num_reemb_var33_ult1', 'num_trasp_var17_out_hace3',
       'num_trasp_var33_out_hace3', 'saldo_var2_ult1',
       'saldo_medio_var13_medio_hace3'],
      dtype='object')

# Quasi Constant Removal

The  rule of thumb, remove those quasi-constant features that have more than 99% similar values for the output observations
<b>value somewhat near to constant</b>

Variance Threshold removes all zero-variance feature i.e feature that have same value in all the samples

In [17]:
#0.1 indicate 99% of observation approximately
Quasi_constant_filter = VarianceThreshold(threshold=0.01)

In [18]:
#fit find the feature with low variance
Quasi_constant_filter.fit(x_train_filter)

VarianceThreshold(threshold=0.01)

get support is a boolean vector that indicates which features are retained. If we
sum over get_support , we get the number of feature that are <b>not</b> Quasi Constant

In [19]:
Quasi_constant_filter.get_support().sum()

267

In [20]:
x_train_quasi_filter = Quasi_constant_filter.transform(x_train_filter)
x_test_quasi_filter = Quasi_constant_filter.transform(x_test_filter)

In [21]:
x_train_quasi_filter.shape, x_test_quasi_filter.shape, x_train_filter.shape

((60816, 267), (15204, 267), (60816, 332))

# Duplicate removal

In [22]:
# Transpose row to column and vice versa

x_train_T = x_train_quasi_filter.T
x_test_T = x_test_quasi_filter.T

In [23]:
# set this into a dataframe

x_train_T =pd.DataFrame(x_train_T)
x_test_T= pd.DataFrame(x_test_T)

In [24]:
x_train_T.shape, x_test_T.shape

((267, 60816), (267, 15204))

In [25]:
# check the number of duplicate values

x_train_T.duplicated().sum()

16

In [26]:
duplicated_features = x_train_T.duplicated()
duplicated_features

0      False
1      False
2      False
3      False
4      False
       ...  
262    False
263    False
264    False
265    False
266    False
Length: 267, dtype: bool

In [28]:
# features which are not duplicated (by creating an inverted list)

features_to_keep = [not index for index in duplicated_features]

In [29]:
x_train_unique = x_train_T[features_to_keep].T
x_test_unique = x_test_T[features_to_keep].T

In [31]:
x_train_unique.shape

(60816, 251)

In [32]:
x_test_unique.shape

(15204, 251)

The original dataset had 370 rows & after removing the duplicates we are left with 251 rows. 

As you can see we have used the below to come to a conclusion as to how many rows we could manage to remove -

1. Constant Removal
2. Quasi Constant Removal
3. Duplicate removal

In [33]:
370-251

119

hence we can conclude that a total number of 119 rows are being removed